Add admin API to get some information about federation status (#11407)
matrix-org/synapse: docs/usage/administration/admin_api/federation.md

- `limit` - Maximum amount of destinations to return. Defaults to `100`.
- `order_by` - The method in which to sort the returned list of destinations.
  Valid values are:
  - `destination` - Destinations are ordered alphabetically by remote server name.
    This is the default.
  - `retry_last_ts` - Destinations are ordered by time of last retry attempt in ms.
  - `retry_interval` - Destinations are ordered by how long until next retry in ms.
  - `failure_ts` - Destinations are ordered by when the server started failing in ms.
  - `last_successful_stream_ordering` - Destinations are ordered by the stream ordering
    of the most recent successfully-sent PDU.
- `dir` - Direction of room order. Either `f` for forwards or `b` for backwards. Setting
  this value to `b` will reverse the above sort order. Defaults to `f`.

*Caution:* The database only has an index on the column `destination`.
This means that if a different sort order is used,
this can cause a large load on the database, especially for large environments.

**Response**

The following fields are returned in the JSON response body:

- `destinations` - An array of objects, each containing information about a destination.

>   Destinations objects contain the following fields:

Reviewer: ```suggestion
  Destination objects contain the following fields:
```

  - `destination` - string - Name of the remote server to federate.

>   - `retry_last_ts` - integer - The last time Synapse tried and failed to reach the
>     remote server, in ms. This is `0` if no further retrying occuring.

Author: ```suggestion
  - `retry_last_ts` - integer - The last time Synapse tried and failed to reach the
    remote server, in ms. This is `0` if the last attempt to communicate with the
    remote server was successful.
```

  - `retry_interval` - integer - How long since the last time Synapse tried to reach
    the remote server before trying again, in ms. This is `0` if no further retrying occuring.

>   - `failure_ts` - integer - The first time Synapse tried and failed to reach the

Author: ```suggestion
  - `failure_ts` - nullable integer - The first time Synapse tried and failed to reach the
```

>   - `failure_ts` - integer - The first time Synapse tried and failed to reach the
>     remote server, in ms. This is `null` if no error has occurred.

Author: ```suggestion
  - `failure_ts` - nullable integer - The first time Synapse tried and failed to reach the
    remote server, in ms. This is `null` if communication with the remote server has never failed.
```

  - `last_successful_stream_ordering` - integer - The stream ordering of the most

>     recent successfully-sent [PDU](Understanding-Synapse-Performance-Issues-Through-Grafana-Graphs.md#federation)

Author: ```suggestion
    recent successfully-sent [PDU](understanding_synapse_through_grafana_graphs.md#federation)
```

>     to this destination or `null` if this information has not been tracked yet.

Reviewer: ```suggestion
    to this destination, or `null` if this information has not been tracked yet.
```

# Federation API

> This API allows a server administrator to manage the federation.

Reviewer: ```suggestion
This API allows a server administrator to manage Synapse's federation with other homeservers.
```

> **Note**: This API is new, experimental and "subject to change".

Reviewer: Not sure the quotes make sense here---they look like [scare quotes](https://en.wikipedia.org/wiki/Scare_quotes) to me, but I'm not sure what you refer to.

## List of destinations

This API gets the current destination retry timing info for a remote server.

The API is:

A standard request with no filtering:

```
GET /_synapse/admin/v1/federation/destinations
```

> A response body like the following is returned:

Reviewer: Might be worth clarifying that a response body is not required, and is ignored?

Author: I am do not understand you. Does you mean an request body? IMO the `GET` has no request body.
Or do you mean, that `destinations` is empty if the server has no federation?

Reviewer: > Does you mean an request body? IMO the GET has no request body.

Yes, that's what I meant. I hadn't considered that this was a GET.

For completeness, quoting [here](https://httpwg.org/specs/rfc7231.html#rfc.section.4.3.1):

> A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request.

```json
{
   "destinations":[
      {
         "destination": "matrix.org",
         "retry_last_ts": 1557332397936,
         "retry_interval": 3000000,
         "failure_ts": 1557329397936,
         "last_successful_stream_ordering": null
      }
   ],
   "total": 1
}
```

To paginate, check for `next_token` and if present, call the endpoint again
with `from` set to the value of `next_token`. This will return a new page.

If the endpoint does not return a `next_token` then there are no more destinations
to paginate through.

**Parameters**

The following query parameters are available:

- `from` - Offset in the returned list. Defaults to `0`.
- `limit` - Maximum amount of destinations to return. Defaults to `100`.
- `order_by` - The method in which to sort the returned list of destinations.
  Valid values are:
  - `destination` - Destinations are ordered alphabetically by remote server name.

>     This is the default.

Reviewer 2: This doesn't sound very stable to me to use as the initial sorting / pagination scheme. This might not matter usually, but is something to note.

Author: The reason for this default sorting is the performances / indexes. All other columns have no indexes. The idea is not to generate a high load by default.

  - `retry_last_ts` - Destinations are ordered by time of last retry attempt in ms.
  - `retry_interval` - Destinations are ordered by how long until next retry in ms.
  - `failure_ts` - Destinations are ordered by when the server started failing in ms.
  - `last_successful_stream_ordering` - Destinations are ordered by the stream ordering
    of the most recent successfully-sent PDU.
- `dir` - Direction of room order. Either `f` for forwards or `b` for backwards. Setting
  this value to `b` will reverse the above sort order. Defaults to `f`.

*Caution:* The database only has an index on the column `destination`.
This means that if a different sort order is used,
this can cause a large load on the database, especially for large environments.

**Response**

> The following fields are returned in the JSON response body:

Reviewer: A couple of questions:

- What happens if I'm federating with a server and federation has never failed? Does it appear in this list?
- What happens if federation with a problematic server recovers---does it appear in this list?
- Does the data returned persist across Synapse restarts?
- I think someone told me that Synapse has some kind of "blocking" mechanism, where if a remote homeserver fails too many federation requests, we block all federation requests to that server. Is that reflected in this output at all?

Sorry---I appreciate that some of these questions are about how Synapse itself works, rather than this API. But I think they're the kind of thing that'd be useful to consider in the documentation.

Author: A few answers:
> - What happens if I'm federating with a server and federation has never failed? Does it appear in this list?

Every federated server should be in the list.
https://github.com/matrix-org/synapse/blob/2b82ec425fccb0ef626242779f7ccd4d77a0685c/synapse/storage/databases/main/transactions.py#L368-L369

> What happens if federation with a problematic server recovers---does it appear in this list?

Yes.
https://github.com/matrix-org/synapse/blob/2b82ec425fccb0ef626242779f7ccd4d77a0685c/synapse/storage/databases/main/transactions.py#L203-L204

> Does the data returned persist across Synapse restarts?

Yes.

> I think someone told me that Synapse has some kind of "blocking" mechanism, where if a remote homeserver fails too many federation requests, we block all federation requests to that server. Is that reflected in this output at all?

I don't know.

- `destinations` - An array of objects, each containing information about a destination.
  Destinations objects contain the following fields:
  - `destination` - string - Name of the remote server to federate.
  - `retry_last_ts` - integer - The last time Synapse tried and failed to reach the

>     remote server, in ms

Reviewer: ```suggestion
    remote server, in ms.
```

  Destinations objects contain the following fields:
  - `destination` - string - Name of the remote server to federate.
  - `retry_last_ts` - integer - The last time Synapse tried and failed to reach the

>     remote server, in ms. This is `0` if no further retrying occuring.

Reviewer: ```suggestion
  - `retry_last_ts` - integer - The last time Synapse tried and failed to reach the
    remote server, in ms. This is `0` if the last attempt to communicate with the remote server was successful.
```

  - `retry_interval` - integer - How long since the last time Synapse tried to reach
    the remote server before trying again, in ms. This is `0` if no further retrying occuring.
  - `failure_ts` - integer - The first time Synapse tried and failed to reach the

>     remote server, in ms. This is `null` if no error has occurred.

Reviewer: ```suggestion
  - `failure_ts` - optional integer - The first time Synapse tried and failed to reach the
    remote server, in ms. This is `null` if communication with the remote server has never failed.
```

Reviewer: Actually, maybe it's better to say "nullable integer", because the field will always be present? Or maybe we just say "integer" elsewhere in the docs?

Author: This is the content of the database:
```
select * from destinations;
 destination | retry_last_ts | retry_interval | failure_ts | last_successful_stream_ordering
-------------+---------------+----------------+------------+---------------------------------
 matrix.org  |             0 |              0 |            |                           29083
```

The result should be:
```json
      {
         "destination": "matrix.org",
         "retry_last_ts": 0,
         "retry_interval": 0,
         "failure_ts": null,
         "last_successful_stream_ordering": 1234
      }
```
